types/wlr_xdg_shell.c

```diff
@@ -131,6 +131,24 @@ static const struct wlr_keyboard_grab_interface xdg_keyboard_grab_impl = {
 	.cancel = xdg_keyboard_grab_cancel,
 };
 
+static void xdg_surface_destroy(struct wlr_xdg_surface *surface);
+
+static void wlr_xdg_popup_grab_handle_seat_destroy(
+		struct wl_listener *listener, void *data) {
+	struct wlr_xdg_popup_grab *xdg_grab =
+		wl_container_of(listener, xdg_grab, seat_destroy);
+
+	wl_list_remove(&xdg_grab->seat_destroy.link);
+
+	struct wlr_xdg_popup *popup, *next;
+	wl_list_for_each_safe(popup, next, &xdg_grab->popups, grab_link) {
+		xdg_surface_destroy(popup->base);
+	}
+
+	wl_list_remove(&xdg_grab->link);
+	free(xdg_grab);
```

Comment:
This event is the only place I see where `xdg_grab` is freed, but there are other places where the grab could be released. I think we are leaking grabs here.

The grab can be released when the grab ends. There is a seat event for this that you can try to hook into.

This should be refactored into a destroy func that runs whenever the grab can be released.

Reply:
I've had a fresh look, I don't think we ever remove the popup from `shell->popup_grabs` (the pendant of `xdg_grab->link`), so we never make the popup unreachable.

Basically, when a grab ends we keep the item in the list and reuse it next time someone wants a grab from the seat, so it's only ever required to free it when the seat is destroyed. If we destroy it all the time we'll need to reallocate it all the time, which I assume is why it was designed that way.

Reply:
Oh I remember that now. I think what you say is right.